uring/src/main/scala/fs2/io/uring/net/UringDatagramSocket.scala

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2022 Arman Bilge
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2
+package io
+package uring
+package net
+
+import cats.effect.kernel.Async
+import cats.effect.kernel.Resource
+import cats.effect.kernel.Sync
+import cats.effect.std.Mutex
+import cats.syntax.all._
+import com.comcast.ip4s.IpAddress
+import com.comcast.ip4s.SocketAddress
+import fs2.Pipe
+import fs2.io.net.Datagram
+import fs2.io.net.DatagramSocket
+import fs2.io.uring.unsafe.uring._
+import fs2.io.uring.unsafe.util._
+
+import scala.scalanative.libc.errno._
+import scala.scalanative.posix.sys.socket._
+import scala.scalanative.posix.sys.socketOps._
+import scala.scalanative.posix.sys.uio
+// import scala.scalanative.posix.sys.uioOps._ Implementation of iovecOps is missing
+import scala.scalanative.unsafe._
+import scala.scalanative.unsigned._
+import com.comcast.ip4s.MulticastJoin
+
+private[net] final class UringDatagramSocket[F[_]](
+    ring: Uring[F],
+    fd: Int,
+    buffer: Ptr[Byte],
+    defaultReadSize: Int,
+    readMutex: Mutex[F],
+    writeMutex: Mutex[F],
+    iov: Ptr[uio.iovec],
+    msg: Ptr[msghdr],
+    remoteAddr: Ptr[sockaddr]
+)(implicit F: Async[F])
+    extends DatagramSocket[F] {
+
+  private[this] def recvmsg(msg: Ptr[msghdr], flags: Int): F[Int] =
+    ring.call(io_uring_prep_recvmsg(_, fd, msg, flags))
+
+  def read: F[Datagram] =
+    readMutex.lock.surround {
+      for {
+        _ <- F.delay {
+          iov._1 = buffer
+          iov._2 = defaultReadSize.toULong
+        }
+
+        _ <- F.delay {
+          msg.msg_name = remoteAddr.asInstanceOf[Ptr[Byte]]
+          msg.msg_namelen = sizeof[sockaddr].toUInt
+          msg.msg_iov = iov
+          msg.msg_iovlen = 1
+          msg.msg_control = null
+          msg.msg_controllen = 0.toUInt
+          msg.msg_flags = 0
+        }
+
+        readed <- recvmsg(msg, 0)
+
+        remoteAddress <- F.fromEither(
+          SocketAddressHelpers
+            .toSocketAddress(remoteAddr)
+        )
```

Comment:
Sorry, I think I made a mistake before. I think this does need to be suspended in a `delay` block, because even though the conversion is not a side-effect, `remoteAddr` is mutable memory and reading mutable memory is a side-effect.